File: awscloud/aws-modern-application-workshop/python/aws-modern-application-workshop/module-5/app/streaming/streamProcessor.py
# The code to be used as an AWS Lambda function for processing real-time
# user click records from Kinesis Firehose and adding additional attributes
# to them before they are stored in Amazon S3.
from __future__ import print_function

import base64

import json

import requests

# Send a request to the Mysfits Service API that we have created in previous
# modules to retrieve all of the attributes for the included MysfitId.
def retrieveMysfit(mysfitId):
    apiEndpoint = 'REPLACE_ME_API_ENDPOINT' + '/mysfits/' + str(mysfitId) # eg: 'https://ljqomqjzbf.execute-api.us-east-1.amazonaws.com/prod/'
    mysfit = requests.get(apiEndpoint).json()
    return mysfit

# The below method will serve as the "handler" for the Lambda function. The
# handler is the method that AWS Lambda will invoke with events, which in this
# case will include records from the Kinesis Firehose Delivery Stream.
def processRecord(event, context):
    output = []

    # retrieve the list of records included with the event and loop through
    # them to retrieve the full list of mysfit attributes and add the additional
    # attributes that a hypothetical BI/Analyitcs team would like to analyze.
    for record in event['records']:
        print('Processing record: ' + record['recordId'])
        # kinesis firehose expects record payloads to be sent as encoded strings,
        # so we must decode the data first to retrieve the click record.
        click = json.loads(base64.b64decode(record['data']))

        mysfitId = click['mysfitId']
        mysfit = retrieveMysfit(mysfitId)

        enrichedClick = {
                'userId': click['userId'],
                'mysfitId': mysfitId,
                'goodevil': mysfit['goodevil'],
                'lawchaos': mysfit['lawchaos'],
                'species': mysfit['species']
            }

       # create the output record that Kinesis Firehose will store in S3.
        output_record = {
            'recordId': record['recordId'],
            'result': 'Ok',
            'data': base64.b64encode(json.dumps(enrichedClick).encode('utf-8') + b'\n').decode('utf-8')
        }
        output.append(output_record)

    print('Successfully processed {} records.'.format(len(event['records'])))

    # return the enriched records to Kiesis Firehose.
    return {'records': output}

File: awscloud/aws-modern-application-workshop/go/aws-modern-application-workshop/module-5/app/streaming/streamProcessor.go
package main

import (
    "github.com/aws/aws-lambda-go/events"
    "github.com/aws/aws-lambda-go/lambda"
    "encoding/json"
    "fmt"
    "net/http"
    "log"
    "io/ioutil"
    "os"
)

// Get Mysfit info from service API
func retrieveMysfit(mysfitId string) string {
    apiEndpoint := "REPLACE_ME_END_POINT"  // eg: https://ljqomqjzbf.execute-api.us-east-1.amazonaws.com/prod/

    resp, err := http.Get(apiEndpoint)
    if err != nil {
        log.Fatalln(err)
    }

    body, err := ioutil.ReadAll(resp.Body)
    if err != nil {
        log.Fatalln(err)
    }

    // Convert body to JSON
    var raw map[string]interface{}
    json.Unmarshal(body, &raw)
    raw["count"] = 1
    out, _ := json.Marshal(raw)

    return string(out)
}

// Lambda function handler
func handleRequest(evnt events.KinesisFirehoseEvent) (events.KinesisFirehoseResponse, error) {
    var response events.KinesisFirehoseResponse

    for _, record := range evnt.Records {
        fmt.Printf("RecordID: %s\n", record.RecordID)
        fmt.Printf("ApproximateArrivalTimestamp: %s\n", record.ApproximateArrivalTimestamp)

        // Transform data: ToUpper the data
        var transformedRecord events.KinesisFirehoseResponseRecord
        transformedRecord.RecordID = record.RecordID
        transformedRecord.Result = events.KinesisFirehoseTransformedStateOk
        // transformedRecord.Data = []byte(strings.ToUpper(string(record.Data)))

        response.Records = append(response.Records, transformedRecord)
    }

    fmt.Println(response)  // Or should this be log.Println???

    return response, nil
}

func main() {
    lambda.Start(handleRequest)
}
